core/backend/src/test/standalone/ChangeConflictHandler.test.ts

```diff
@@ -0,0 +1,369 @@
+    const aspectId2 = insertExternalSourceAspect(b1, el1, "test identifier");
+    b1.saveChanges();
+    await b1.pushChanges({ accessToken: accessToken1, description: "" });
+
+    await b1.pullChanges();
+    await b2.pullChanges();
+
+    updateExternalSourceAspect(b1, aspectId1, el1, "hello1");
+    updateExternalSourceAspect(b2, aspectId2, el1, "hello2");
+
+    b1.saveChanges();
+    b2.saveChanges();
+
+    await spyChangesetConflictHandler(
+      b2,
+      async () => b2.pushChanges({ accessToken: accessToken2, description: "" }),
+      (spy) => expect(spy.callCount).eq(0, "changeset conflict handler should not be called"),
+    );
+
+    await spyChangesetConflictHandler(
+      b1,
+      async () => assertThrowsAsync(
+        async () => b1.pushChanges({ accessToken: accessToken1, description: "" }),
+        "UPDATE/DELETE before value do not match with one in db or CASCADE action was triggered."),
+      (spy) => {
+        expect(spy.callCount).eq(1);
+        expect(spy.alwaysReturned(DbConflictResolution.Abort)).true;
+        const arg = spy.args[0][0];
+        expect(arg.cause).eq(DbConflictCause.Data);
+        expect(arg.opcode).eq(DbOpcode.Update);
+        expect(arg.indirect).false;
+        expect(arg.tableName).eq("bis_Element");
+      },
+    );
+  });
+  it("DbConflictCause.NotFound - deleted row", async () => {
+    const el1 = insertPhysicalObject(b1);
+
+    b1.saveChanges();
+    await b1.pushChanges({ accessToken: accessToken1, description: "" });
+
+    await b1.pullChanges();
+    await b2.pullChanges();
+
+    b1.elements.deleteElement(el1);
+    insertExternalSourceAspect(b2, el1, "test identifier");
+
+    b1.saveChanges();
+    b2.saveChanges();
+
+    await spyChangesetConflictHandler(
+      b1,
+      async () => b1.pushChanges({ accessToken: accessToken1, description: "" }),
+      (spy) => expect(spy.callCount).eq(0, "changeset conflict handler should not be called"),
+    );
+
+    // push will fail with Data conflict and to simulate FK we will for force push by overriding
+    // return result of conflict Handler.
+    await stubChangesetConflictHandler(
+      b2,
+      async () => b2.pushChanges({ accessToken: accessToken1, description: "" }),
+      (stub) => {
+        stub.returns(DbConflictResolution.Skip);
+      },
+    );
+
+    // third briefcase while pull will see a fk violation.
+    await spyChangesetConflictHandler(
+      b3,
+      async () => b3.pullChanges({ accessToken: accessToken1 }),
+      (spy) => {
+        expect(spy.callCount).eq(1);
+        expect(spy.alwaysReturned(DbConflictResolution.Skip)).true;
+        const arg = spy.args[0][0];
+        expect(arg.cause).eq(DbConflictCause.NotFound);
+        expect(arg.opcode).eq(DbOpcode.Delete);
+        expect(arg.indirect).true;
+        expect(arg.tableName).eq("bis_GeometricElement3d");
+      },
+    );
+  });
+  it.skip("DbConflictCause.Constraint - duplicate userLabels", async () => {
```

Comment:
@khanaffan I failed to notice in the original PR - why is this test skipped?

Reply:
Yes, that needs to be skipped as I did not find any solution when a constraint violation happens. When it happens the saveChanges/abandonChanges() both fail. I did not dig into it but we technically cannot close the db.

I will write a issue for it as it need more investigation.

Reply:
Also, this is not a known issue. My effort was to test all cases in a conflict handler and so this test simulated a case. Its not any new issue its just not covered by any test up till now.